### PR TITLE
Model 1.1 Implementation

### PR DESCRIPTION
**Pull request by patricktnast**
### Description

- *Category*:  implementation
- *JIRA issue*: [MIC-4159](https://jira.ihme.washington.edu/browse/MIC-4159)
- *Research reference*:  [Model Table](https://vivarium-research.readthedocs.io/en/latest/models/concept_models/vivarium_nutrition_optimization/pregnancies/concept_model.html)

### Changes and notes
 
Model 1.1 specifies that on initialization, simulants should be assigned a full/partial term length and a corresponding pregnancy duration. both are added into the state table on initialization, but the duration itself is passed into the dwell time pipeline.

### Verification and Testing

Simulation runs; simulants are assigned full / partial terms according to the correct distribution; partial term simulants are assigned term lengths uniformly from 6-24 weeks; full term simulants are assigned 40 week term.

CI passes.

### CODE REVIEW

File: src/vivarium_gates_nutrition_optimization/constants/data_values.py [hunk 1]
```diff
@@ -1,5 +1,15 @@
+from typing import NamedTuple
+
+
 ############################
 # Disease Model Parameters #
 ############################
+class _Durations(NamedTuple):
+     ## Days
+    FULL_TERM = 40 * 7
```
Comment (mattkappel) [inline src/vivarium_gates_nutrition_optimization/constants/data_values.py:9]:
```suggestion
    ## Days
    FULL_TERM = 40 * 7
```

Comment (mattkappel) [reply src/vivarium_gates_nutrition_optimization/constants/data_values.py:9]: Were you intentionally offsetting by 1 column or typo?

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py [hunk 2]
```diff
@@ -16,9 +18,88 @@ def __init__(self, cause, *args, **kwargs):
...
+            if transition.start_active:
+                transition.set_active(pop_data.index)
+
+        pop_events = self.get_initial_event_times(pop_data)
+        pregnancy_term_outcomes_and_durations = self.sample_pregnancy_terms_and_durations(
+            pop_data
+        )
+        pop_update = pd.concat([pop_events, pregnancy_term_outcomes_and_durations], axis=1)
+        self.population_view.update(pop_update)
+
```
Comment (mattkappel) [inline src/vivarium_gates_nutrition_optimization/components/pregnancy.py:55]: I assume, this is working, but I would have expected you need to make a `builder.population.initializes_simulants` call like [here](https://github.com/ihmeuw/vivarium_gates_iv_iron/blob/862b057457cea2410e8d24c2c9c38c43419cd8b2/src/vivarium_gates_iv_iron/components/pregnancy.py#L146-L151). It doesn't look like it happens in the superclass's setup either.

Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/components/pregnancy.py:55]: this happens in the setup for the supersuperclass `BaseDiseaseState`

**Review by mattkappel — APPROVED** [review]

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py [hunk 3]
```diff
@@ -16,9 +18,88 @@ def __init__(self, cause, *args, **kwargs):
...
+                DURATIONS.FULL_TERM, unit="days"
+            ),
+            models.PARTIAL_TERM_OUTCOME: self.sample_partial_term_durations,
+        }
+
+        for term_length, sampling_func in term_duration_map.items():
+            term_pop = pregnancy_term_outcomes[
+                pregnancy_term_outcomes["pregnancy_term_outcome"] == term_length
+            ].index
+            pregnancy_term_outcomes.loc[term_pop, "pregnancy_duration"] = sampling_func(
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization/components/pregnancy.py:86]: Nit: any reason not to call this `sampling_function` rather than `sampling_func`?

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py [hunk 4]
```diff
@@ -16,9 +18,88 @@ def __init__(self, cause, *args, **kwargs):
...
+                term_pop
+            )
+
+        return pregnancy_term_outcomes
+
+    def sample_partial_term_durations(self, partial_term_pop: pd.DataFrame) -> pd.Series:
+        low, high = DURATIONS.DETECTION, DURATIONS.PARTIAL_TERM
+        draw = self.randomness.get_draw(partial_term_pop, additional_key="pregnancy_duration")
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization/components/pregnancy.py:94]: So you don't have to risk a potential key collision when we add sampling for full-term duration, it'd probably be better to call the additional key something like `"partial_term_pregnancy_duration"`

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py [hunk 5]
```diff
@@ -47,7 +127,7 @@ def Pregnancy():
             "prevalence": lambda *_: 0.0,
             "disability_weight": lambda *_: 0.0,
             "excess_mortality_rate": lambda *_: 0.0,
-            "dwell_time": lambda *_: pd.Timedelta(days=POSTPARTUM_DURATION),
+            "dwell_time": lambda *_: pd.Timedelta(days=DURATIONS.POSTPARTUM_DURATION),
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization/components/pregnancy.py:130]: Nit: since the class is already called `DURATIONS`, you can shorten this to `POSTPARTUM`.

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py [hunk 6]
```diff
@@ -61,3 +141,29 @@ def Pregnancy():
...
+
+
+def get_partial_term_probability_data(builder: Builder) -> pd.DataFrame:
+    asfr = builder.data.load(data_keys.PREGNANCY.ASFR).set_index(ARTIFACT_INDEX_COLUMNS)
+    sbr = (
+        builder.data.load(data_keys.PREGNANCY.SBR)
+        .set_index("year_start")
+        .drop(columns=["year_end"])
+        .reindex(asfr.index, level="year_start")
+    )
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization/components/pregnancy.py:153]: What is the structure of SBR in the artifact that means you have to do this reshaping?

File: src/vivarium_gates_nutrition_optimization/constants/data_values.py [hunk 7]
```diff
@@ -1,5 +1,15 @@
...
+class _Durations(NamedTuple):
+    ## Days
+    FULL_TERM = 40 * 7
+    POSTPARTUM_DURATION = 6 * 7
+    DETECTION = 6 * 7
+    PARTIAL_TERM = 24 * 7
+
 
-POSTPARTUM_DURATION = 6 * 7  ## Days
+DURATIONS = _Durations()
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization/constants/data_values.py:15]: Are any of these going to be used long-term besides post-partum?

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_gates_nutrition_optimization/components/pregnancy.py (hunk 6 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/components/pregnancy.py:153]: SBR is not age specific, so it is essentially just a single value indexed by the year for a particular location. This reshaping is intended to extend it onto the relevant multiindex

File: src/vivarium_gates_nutrition_optimization/constants/data_values.py (hunk 7 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/constants/data_values.py:15]: I think PARTIAL_TERM, FULL_TERM, and DETECTION will remain used in a similar context to what they're being used for now, to set the scaling factors for the pregnancy duration draws, even as the full term distribution gets more complicated with e.g. LBWSG

**Review by rmudambi — APPROVED** [review]
